File: sdk/android/causality/src/main/kotlin/io/causality/Causality.kt
package io.causality

import android.app.Application
import android.content.Context
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import io.causality.internal.Bridge
import io.causality.internal.Platform
import kotlinx.coroutines.*

/**
 * Main entry point for the Causality analytics SDK.
 *
 * Usage:
 * ```kotlin
 * // Initialize in Application.onCreate()
 * Causality.initialize(this) {
 *     apiKey = "your-api-key"
 *     endpoint = "https://your-server.com"
 *     appId = "your-app-id"
 * }
 *
 * // Track events
 * Causality.track("button_tap") {
 *     property("button_name", "checkout")
 * }
 *
 * // Or using the Event DSL
 * Causality.track(event("purchase") {
 *     property("product_id", "abc123")
 *     property("price", 29.99)
 * })
 * ```
 */
object Causality : DefaultLifecycleObserver {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var initialized = false

    /**
     * Initialize the SDK with configuration using DSL builder.
     *
     * @param context Application context
     * @param block Configuration builder
     */
    fun initialize(context: Context, block: ConfigBuilder.() -> Unit) {
        initialize(context, config(block))
    }

    /**
     * Initialize the SDK with configuration.
     *
     * @param context Application context
     * @param config SDK configuration
     */
    fun initialize(context: Context, config: Config) {
        // Set platform context first
        Platform.collectAndSetContext(context.applicationContext)

        // Initialize Go core
        Bridge.initSDK(config)
        initialized = true

        // Register lifecycle observer
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    /**
     * Track an event using DSL builder.
     *
     * @param type Event type
     * @param block Event builder
     */
    fun track(type: String, block: EventBuilder.() -> Unit = {}) {
        track(event(type, block))
    }

    /**
     * Track an event.
     *
     * @param event The event to track
     */
    fun track(event: Event) {
        if (!initialized) {
            if (BuildConfig.DEBUG) {
                android.util.Log.w("Causality", "SDK not initialized, event dropped")
            }
            return
        }

        scope.launch {
            try {
                Bridge.track(event)
            } catch (e: Exception) {
                if (BuildConfig.DEBUG) {
                    android.util.Log.e("Causality", "Track error", e)
                }
            }
        }
    }

    /**
     * Set user identity.
     *
     * @param userId Unique user identifier
     * @param traits Optional user properties
     * @param aliases Optional alias identifiers
     */
    fun identify(userId: String, traits: Map<String, String>? = null, aliases: List<String>? = null) {
        if (!initialized) throw CausalityException.NotInitialized()
        Bridge.setUser(userId, traits, aliases)
    }

    /**
     * Clear user identity (soft reset - keeps device ID).
     */
    fun reset() {
        if (!initialized) throw CausalityException.NotInitialized()
        Bridge.reset()
    }

    /**
     * Full reset - clears user identity and regenerates device ID.
     */
    fun resetAll() {
        if (!initialized) throw CausalityException.NotInitialized()
        Bridge.resetAll()
    }

    /**
     * Force flush all queued events.
     */
    suspend fun flush() {
        if (!initialized) throw CausalityException.NotInitialized()
        withContext(Dispatchers.IO) {
            Bridge.flush()
        }
    }

    /**
     * Get the device identifier.
     */
    val deviceId: String
        get() = Bridge.getDeviceId()

    /**
     * Check if SDK is initialized.
     */
    val isInitialized: Boolean
        get() = Bridge.isInitialized()

    // Lifecycle callbacks
    override fun onStop(owner: LifecycleOwner) {
        Bridge.appDidEnterBackground()
    }

    override fun onStart(owner: LifecycleOwner) {
        Bridge.appWillEnterForeground()
    }

    // Convenience methods

    /**
     * Track a screen view.
     */
    fun trackScreenView(name: String, properties: Map<String, Any>? = null) {
        track("screen_view") {
            property("screen_name", name)
            properties?.forEach { (k, v) ->
                when (v) {
                    is String -> property(k, v)
                    is Int -> property(k, v)
                    is Long -> property(k, v)
                    is Double -> property(k, v)
                    is Boolean -> property(k, v)
                }
            }
        }
    }

    /**
     * Track a button tap.
     */
    fun trackButtonTap(name: String, properties: Map<String, Any>? = null) {
        track("button_tap") {
            property("button_name", name)
            properties?.forEach { (k, v) ->
                when (v) {
                    is String -> property(k, v)
                    is Int -> property(k, v)
                    is Long -> property(k, v)
                    is Double -> property(k, v)
                    is Boolean -> property(k, v)
                }
            }
        }
    }
}

File: sdk/ios/Sources/Causality/Causality.swift
import Foundation
import CausalityCore
#if canImport(UIKit)
import UIKit
#endif

/// Main entry point for the Causality analytics SDK
@MainActor
public final class Causality {
    /// Shared singleton instance
    public static let shared = Causality()

    private var isInitialized = false

    private init() {
        // Register for lifecycle notifications
        #if canImport(UIKit)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidEnterBackground),
            name: UIApplication.didEnterBackgroundNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillEnterForeground),
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )
        #endif
    }

    /// Initialize the SDK with configuration
    /// - Parameter config: SDK configuration
    /// - Throws: CausalityError if initialization fails
    public func initialize(config: Config) throws {
        // Set platform context first
        Platform.setPlatformContext()

        // Initialize Go core
        try Bridge.initSDK(config: config)
        isInitialized = true
    }

    /// Track a freeform event
    /// - Parameter event: The event to track
    /// - Note: This method is non-blocking. Events are queued for batch sending.
    public func track(_ event: Event) {
        guard isInitialized else {
            #if DEBUG
            print("[Causality] Warning: SDK not initialized, event dropped")
            #endif
            return
        }

        Task.detached(priority: .utility) {
            do {
                try Bridge.track(event: event)
            } catch {
                #if DEBUG
                print("[Causality] Track error: \(error)")
                #endif
            }
        }
    }

    /// Track a typed event with compile-time safety
    /// - Parameter event: A CausalityEvent (e.g., ScreenView, ButtonTap, PurchaseComplete)
    /// - Note: This method is non-blocking. Events are queued for batch sending.
    public func track<E: CausalityEvent>(_ event: E) {
        guard isInitialized else {
            #if DEBUG
            print("[Causality] Warning: SDK not initialized, event dropped")
            #endif
            return
        }

        Task.detached(priority: .utility) {
            do {
                try Bridge.trackTyped(event: event)
            } catch {
                #if DEBUG
                print("[Causality] Track error: \(error)")
                #endif
            }
        }
    }

    /// Track an event using a builder
    /// - Parameter type: Event type
    /// - Returns: EventBuilder for fluent property addition
    public func track(type: String) -> EventBuilder {
        EventBuilder(type: type)
    }

    /// Set user identity
    /// - Parameters:
    ///   - userId: Unique user identifier
    ///   - traits: Optional user properties
    ///   - aliases: Optional alias identifiers
    public func identify(userId: String, traits: [String: AnyCodable]? = nil, aliases: [String]? = nil) throws {
        guard isInitialized else {
            throw CausalityError.notInitialized
        }
        try Bridge.setUser(userId: userId, traits: traits, aliases: aliases)
    }

    /// Clear user identity (soft reset - keeps device ID)
    public func reset() throws {
        guard isInitialized else {
            throw CausalityError.notInitialized
        }
        try Bridge.reset()
    }

    /// Full reset - clears user identity and regenerates device ID
    public func resetAll() throws {
        guard isInitialized else {
            throw CausalityError.notInitialized
        }
        try Bridge.resetAll()
    }

    /// Force flush all queued events
    public func flush() async throws {
        guard isInitialized else {
            throw CausalityError.notInitialized
        }

        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            DispatchQueue.global(qos: .utility).async {
                do {
                    try Bridge.flush()
                    continuation.resume()
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Get the device identifier
    /// - Returns: Device ID, or empty string if not initialized
    public var deviceId: String {
        Bridge.getDeviceId()
    }

    /// Check if SDK is initialized
    public var initialized: Bool {
        Bridge.isInitialized()
    }

    // MARK: - Lifecycle

    @objc private func appDidEnterBackground() {
        Bridge.appDidEnterBackground()
    }

    @objc private func appWillEnterForeground() {
        Bridge.appWillEnterForeground()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

// MARK: - Convenience extensions

public extension Causality {
    /// Track a screen view
    func trackScreenView(name: String, screenClass: String? = nil, previousScreen: String? = nil) {
        track(ScreenView(screenName: name, screenClass: screenClass, previousScreen: previousScreen))
    }

    /// Track a button tap
    func trackButtonTap(id: String, text: String? = nil, screenName: String? = nil) {
        track(ButtonTap(buttonId: id, buttonText: text, screenName: screenName))
    }
}
